functions/src/index.ts

```diff
@@ -206,3 +107,172 @@ function roundDate(d: Date) {
   }
   return date;
 }
+
+function cycleDateData(docs: any, separateDates: any[], dates: any) {
+  let beginTime = 24 * 60;
+  let endTime = 0;
+  for (const d of dates) {
+    const dateData = docs.filter((doc: any) => {
+      const t = moment(doc.get("time").toDate());
+      return d.isSame(t, "day");
+    });
+    let data = [];
+    if (dateData.length !== 0) {
+      const timeData = dateData.map((doc: any) =>
+        moment(doc.get("time").toDate())
+      );
+      const earliest = moment.min(timeData);
+      const earliestTime = earliest.hour() * 60 + earliest.minute();
+      beginTime = Math.min(beginTime, earliestTime);
+      const latest = moment.max(timeData);
+      const latestTime = latest.hour() * 60 + latest.minute();
+      endTime = Math.max(endTime, latestTime);
+      data = dateData;
+    }
+    separateDates.push(data);
+  }
+  return [beginTime, endTime];
+}
+
+function generateDates(startDate: any, endDate: any) {
+  const dates = [];
+  const currDate = startDate.clone();
+  while (currDate.diff(endDate) < 0) {
+    dates.push(currDate.clone());
+    currDate.add(1, "days");
+  }
+  return dates;
+}
+
+function generateTimes(begin: any, end: any) {
+  const times = [];
+  const currentTime = moment.duration(roundTime(begin), "minutes");
+  let endTime = moment.duration(roundTime(end), "minutes");
+  while (currentTime <= endTime) {
+    times.push(currentTime.clone());
+    currentTime.add(round30 ? 30 : 15, "minutes");
+  }
+  return times;
+}
+
+function addCell(separateDate: any, time: any, rows: string[][]) {
+  const dateData = separateDate.filter((doc: any) => {
+    const t = moment(doc.get("time").toDate());
+    const roundedTime = roundDate(t);
+    const timeInMinutes = roundedTime.hour() * 60 + roundedTime.minute();
+    const timeDuration = moment.duration(timeInMinutes, "minutes");
+    return time.as("minutes") === timeDuration.as("minutes");
+  });
+
+  let cardioGranularData = "";
+  let weightsGranularData = "";
+  let cardioTotalData = "";
+  let weightsTotalData = "";
+  if (dateData.length !== 0) {
+    const doc = dateData[dateData.length - 1]; // latest data entry
+    const cardioDoc = doc.get("cardio");
+    const treadmills = cardioDoc.treadmills;
+    const ellipticals = cardioDoc.ellipticals;
+    const bikes = cardioDoc.bikes;
+    const amts = cardioDoc.amts;
+    const cardioGranularCell =
+      "Treadmills: " +
+      treadmills +
+      " \nEllipticals: " +
+      ellipticals +
+      " \nBikes: " +
+      bikes +
+      " \nAMTs: " +
+      amts;
+    const cardioTotalCell = treadmills + ellipticals + bikes + amts;
+    cardioGranularData = cardioGranularCell;
+    cardioTotalData = cardioTotalCell;
```

Comment:
Is there a reason why you create a new `const` instead of assigning directly to `cardioGranularData`?

Reply:
Hmm not that I can remember, I was probably preparing for more manipulation. I'll simplify it.